hazelcast-client-legacy/src/main/java/com/hazelcast/client/cache/impl/OneShotExecutionCallback.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.cache.impl;
+
+import com.hazelcast.core.ExecutionCallback;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+abstract class OneShotExecutionCallback<V> implements ExecutionCallback<V> {
+
+    private static final int NOT_CALLED = 0;
+    private static final int CALL_IN_PROGRESS = 1;
+    private static final int CALL_FINISHED = 2;
+
+    private final AtomicInteger callState = new AtomicInteger(0);
```

Comment:
yes, `AtomicLongFieldUpdater` makes sense because we may create many `OneShotExecutionCallback`